src/workerd/api/sockets.c++

```diff
@@ -60,122 +69,23 @@ jsg::Ref<Socket> connectImpl(
-  thenOrRunNow<void>([](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->abortRead();
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::getsockopt(int level, int option, void* value, uint* length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->getsockopt(level, option, value, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::setsockopt(int level, int option, const void* value, uint length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->setsockopt(level, option, value, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::getsockname(struct sockaddr* addr, uint* length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->getsockname(addr, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-void PipelinedAsyncIoStream::getpeername(struct sockaddr* addr, uint* length) {
-  thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    (*stream)->getpeername(addr, length);
-    return kj::READY_NOW;
-  }).detach([this](kj::Exception&& exception) mutable {
-    error = kj::mv(exception);
-  });
-}
-
-kj::Promise<size_t> PipelinedAsyncIoStream::read(void* buffer, size_t minBytes, size_t maxBytes) {
-  return thenOrRunNow<size_t>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->read(buffer, minBytes, maxBytes);
-  });
-}
-
-kj::Promise<size_t> PipelinedAsyncIoStream::tryRead(void* buffer, size_t minBytes, size_t maxBytes) {
-  return thenOrRunNow<size_t>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->tryRead(buffer, minBytes, maxBytes);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::write(const void* buffer, size_t size) {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->write(buffer, size);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->write(pieces);
-  });
-}
-
-kj::Promise<void> PipelinedAsyncIoStream::whenWriteDisconnected() {
-  return thenOrRunNow<void>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->whenWriteDisconnected();
-  });
+  // Forcibly close the readable/writable streams.
+  auto& context = IoContext::current();
+  context.addTask(context.run([this](jsg::Lock& js) {
+    readable->getController().cancel(js, nullptr).then(js, [this](
+        jsg::Lock& js) mutable {
+      writable->getController().abort(js, nullptr).then(js,
```

Comment:
Any particular reason `redable.cancel()` and `writable.abort()` have to be in series rather than parallel?

Reply:
No reason. I just figured it was marginally better to try cancelling one of them and not bother cancelling the other if the first cancel operation fails. Any reason doing this in parallel would be better?

Reply:
Well, it seems to me that after `close()` is called, we'd want any methods called on `writable` to immediately behave as if `abort()` had been called on it. But with this implementation, `abort()` doesn't actually get called until a while later.

That said, it's not clear to me why `cancel()` and `abort()` aren't synchronous. When would these calls ever need to wait for I/O, just to cancel it? I guess that's a question for the streams authors, though.

Reply:
Relatedly, I feel like `close()` should not return a promise, but for the same reason I think `cancel()` and `abort()` shouldn't. But `cancel()` and `abort()` _do_ return promises, so perhaps `close()` should as well. I suppose it should return essentially `Promise.all(readable.cancel(), writable.abort())`.

Reply:
Hm, actually now that I tried to refactor this I realised the real reason I did it this way: it's easier to resolve the `closed` promise successfully once I know that both readable and writable have been closed. If I call them in parallel then it becomes tricky, because when do I resolve the promise? I need to know when both have been closed successfully so I might as well close the synchronously. Note that the `cancel` and `abort` operations can fail, and their failure causes the `closed` promise to be resolved with an error.

Hope that makes sense and I didn't miss anything.

Reply:
Returning a Promise.all from close should work, I'd think?

Reply:
Chatted with James about this: TL;DR added a TODO for this.